File: dotnet/src/PuzzleSolver.NumbersGame/SolverResult.cs
namespace PuzzleSolver.NumbersGame;

public class SolverResult
{
    private readonly Board _board;
    private readonly Target _target;

    public Board Board => _board;
    
    public Target Target => _target;

    public bool SolutionFound { get; init; }

    public IReadOnlyCollection<SolveInstruction> Instructions { get; init; }

    internal SolverResult(Board board, Target target)
        : this(board, target, [])
    {
    }

    internal SolverResult(Board board, Target target, Solution solution)
        : this(board, target, [ solution ])
    {
    }

    private SolverResult(Board board, Target target, IEnumerable<Solution> solutions)
    {
        _board = board;
        _target = target;
        var solution = solutions.FirstOrDefault();
        SolutionFound = solution is not null;
        
        var instructions = new List<SolveInstruction>();

        if (SolutionFound)
        {
            instructions.Add(new InitialSolveInstruction(solution!.Start));
            var finalState = solution.Start;

            foreach (var step in solution.Steps)
            {
                instructions.Add(new AdditionalSolveInstruction(step.From, step.Operation, step.Result));
                finalState = step.Result;
            }

            instructions.Add(new FinalSolveInstruction(finalState));
        }

        Instructions = instructions.AsReadOnly();
    }
}

public abstract record SolveInstruction(Board State);

public sealed record InitialSolveInstruction(Board State) : SolveInstruction(State);

public sealed record AdditionalSolveInstruction(Board State, MathematicalOperation Operation, Board Result) : SolveInstruction(State);

public sealed record FinalSolveInstruction(Board State) : SolveInstruction(State);

File: python/solver_result.py
from typing import Self

from board import Board
from solution import Solution
from solve_instruction import (
    AdditionalSolveInstruction,
    FinalSolveInstruction,
    InitialSolveInstruction,
    SolveInstruction,
)
from target import Target


class SolverResult:
    @property
    def board(self) -> Board:
        return self._board

    @property
    def target(self) -> Target:
        return self._target

    @property
    def solution_found(self) -> bool:
        return self._solution_found

    @property
    def instructions(self) -> list[SolveInstruction]:
        return self._instructions

    def __init__(self, board: Board, target: Target, solutions: list[Solution]) -> None:
        """
        Initializes the SolverResult with the board, target, and solutions.

        :param board: The board containing numbers.
        :param target: The target number to reach.
        :param solutions: A list of solutions found to reach the target.
        """

        self._board = board
        self._target = target
        solution = solutions[0] if solutions else None
        self._solution_found = solution is not None

        self._instructions: list[SolveInstruction] = []

        if solution is not None:
            self._instructions.append(InitialSolveInstruction(solution.start))
            final_state = solution.start

            for step in solution.steps:
                self._instructions.append(
                    AdditionalSolveInstruction(
                        step.result,
                        step.operation,
                        step.source))
                final_state = step.result

            self._instructions.append(FinalSolveInstruction(final_state))

    @classmethod
    def unsolvable(cls, board: Board, target: Target) -> Self:
        """
        Factory method to create an unsolvable SolverResult.

        :param board: The board containing numbers.
        :param target: The target number to reach.
        :return: An instance of SolverResult indicating no solutions found.
        """
        return cls(board, target, [])

    @classmethod
    def with_solutions(
        cls, board: Board, target: Target, solutions: list[Solution]
    ) -> Self:
        """
        Factory method to create a SolverResult with solutions.

        :param board: The board containing numbers.
        :param target: The target number to reach.
        :param solutions: A list of solutions found to reach the target.
        :return: An instance of SolverResult with the provided solutions.
        """
        return cls(board, target, solutions)
